### ISSUE TITLE
Are all Multi-Head Attention "Masked" really?

### ISSUE THREAD

**Issue by kuraga** [harvardnlp/annotated-transformer#125]
On the Figure, there `Multi-Head Attention`s and `Masked Multi-Head Attention`s.
Are all Multi-Head Attention "Masked" really?

![image](https://github.com/user-attachments/assets/02b878a5-36ad-4d4f-8a34-51efe3084202)

**Comment by PangLuo** [comment]
Multi-head attention uses masks to prevent attention computations over padded positions.

Masked Multi-Head Attention uses masks in two ways:
* prevent attention computations over padded positions.
* prevent positions from attending to subsequent positions

**Comment by kuraga** [comment]
@PangLuo , thanks for the reply!

> Multi-head attention uses masks to prevent attention computations over padded positions.
> 
> Masked Multi-Head Attention uses masks in two ways:
> 
> * prevent attention computations over padded positions.
> * prevent positions from attending to subsequent positions

![image](https://github.com/user-attachments/assets/81349f7b-a17a-4ec7-83af-63f1a3dbd90e)

So, the lower (first) block should differ from the upper 9second) block with:

> * prevent positions from attending to subsequent positions

The code:

```python
class DecoderLayer(nn.Module):
    <...>

    def forward(self, x, memory, src_mask, tgt_mask):
        "Follow Figure 1 (right) for connections."
        m = memory
        x = self.sublayer[0](x, lambda x: self.self_attn(x, x, x, tgt_mask))
        x = self.sublayer[1](x, lambda x: self.src_attn(x, m, m, src_mask))
        return self.sublayer[2](x, self.feed_forward)
```

So, is this difference - `tgt_mask` and `src_mask`, respectively?

Thanks!

**Comment by PangLuo** [comment]
@kuraga 
You can also see clearly how the authors constructed tgt_mask and src_mask here:

![image](https://github.com/user-attachments/assets/fa3a8c2b-d6d7-4081-85c8-be0d99c3cfa5)

**Comment by kuraga** [comment]
@PangLuo , thanks!

But is this statement true?

> The difference (in this repository) between
> `Masked Multi-Head Attention` (i.e. the lower block, i.e. the first code line)
> and `Multi-Head Attention` (i.e. the upper block, i.e. the second code line)
> is the replacement `tgt_mask` by `src_mask`.

**Comment by PangLuo** [comment]
@kuraga 
> The difference (in this repository) between
Masked Multi-Head Attention (i.e. the lower block, i.e. the first code line)
and Multi-Head Attention (i.e. the upper block, i.e. the second code line)
is the replacement tgt_mask by src_mask.

That's one of the differences. There are two lines in the definition of the class DecoderLayer.
The first line corresponds to the Masked Multi-Head Attention and all of the queries/keys/values
are the decoder input. The calculated scores are masked with tgt_mask.
The second line corresponds to the Multi-Head Attention and the queries are the decoder input, while the keys and values (the two m's) are the memory from the encoder. The calculated scores are masked with src_mask.

x = self.sublayer[0](x, lambda x: self.self_attn(x, x, x, tgt_mask))
x = self.sublayer[1](x, lambda x: self.src_attn(x, m, m, src_mask))

**Comment by kuraga** [comment]
Ok, thanks!

**Comment by PangLuo** [comment]
no worries. Please note I've updated my reply.
